### PR TITLE
Updates based on feedback on #1222 and more

### PR DESCRIPTION
**Pull request by AnnikaH**
- Updates based on feedback on pull request #1222 
- Socket moved out of namespace tcp
- Error handling with net::Error (and net::ICMP_error)
- Handling of Error callbacks in UDP
- DNS integration test
and more

### CODE REVIEW

File: api/net/ip4/udp.hpp [hunk 1]
```diff
@@ -141,27 +150,52 @@ namespace net {
...
-    port_t      current_port_;
+    port_t                      current_port_;
 
     // the async send queue
     std::deque<WriteBuffer> sendq;
+
+    /** Error entries are just error callbacks and timestamps */
+    class Error_entry {
+    public:
+      Error_entry(UDP::error_handler cb) noexcept
```
Comment (AndreasAakesson) [inline api/net/ip4/udp.hpp]: By moving the implementation of the constructor to the cpp you can avoid including `<rtc>` in the header (reducing the scope).

File: api/net/ip4/udp.hpp [hunk 2]
```diff
@@ -141,27 +150,52 @@ namespace net {
...
+
+      UDP::error_handler callback;
+
+    private:
+      RTC::timestamp_t timestamp;
+
+    }; //< class Error_entry
+
+    /** The error callbacks that the user has sent in via the UDPSockets' sendto and bcast methods */
+    std::unordered_map<Socket, Error_entry, Socket::pair_hash> error_callbacks_;
```
Comment (AndreasAakesson) [inline api/net/ip4/udp.hpp]: A implementation of `std::hash<Socket>` would make it possible to drop the third template arg `Socket::pair_hash` (this function can be called inside the std::hash`). See http://en.cppreference.com/w/cpp/utility/hash

File: api/net/socket.hpp [hunk 3]
```diff
@@ -138,12 +147,37 @@ class Socket {
...
 
-} //< namespace tcp
+class Quadruple {
+
+public:
+  Quadruple() noexcept
+    : source_{}, destination_{}
+  {}
+
+  Quadruple(const Socket::Address src_address, const Socket::port_t src_port,
```
Comment (AndreasAakesson) [inline api/net/socket.hpp]: Maybe add additional constructor `Quadruple(Socket, Socket)`.

File: api/net/ip4/udp.hpp (hunk 2 above)
Comment (AnnikaH) [inline api/net/ip4/udp.hpp]: 👍

File: api/net/socket.hpp (hunk 3 above)
Comment (AnnikaH) [inline api/net/socket.hpp]: 👍

File: api/net/inet.hpp [hunk 4]
```diff
@@ -44,28 +44,28 @@ namespace net {
     using Route_checker = delegate<bool(typename IPV::addr)>;
     using IP_packet_factory = delegate<typename IPV::IP_packet_ptr(Protocol)>;
 
-    using Error_type = icmp4::Type;
-    using Error_code = uint8_t;
-
     template <typename IPv>
-    using resolve_func = delegate<void(typename IPv::addr)>;
+    using resolve_func = delegate<void(typename IPv::addr, Error&)>;
```
Comment (AnnikaH) [inline api/net/inet.hpp]: No seems like Error& can be const here =)

File: api/net/ip4/udp.hpp (hunk 1 above)
Comment (AnnikaH) [inline api/net/ip4/udp.hpp]: private attribute RTC::timestamp_t, but could consider int64_t

**Comment by alfred-bratterud** [comment]
jenkins test please

**Comment by mnordsletten** [comment]
This is the error I was seeing yesterday as well @AnnikaH where it seems like you were not expecting the 10.0.0.1 dns server to actually give a valid response. (but it seems like it does)

`Resolved IP address of hotmail.com with DNS server 10.0.0.1: 157.56.198.220`

**Pull request merged**